### PR TITLE
IGNITE-16268 JDBC. Validation of statement type should be done before its execution

### PR DESCRIPTION
**Pull request by vladErmakov07**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 1]
```diff
@@ -230,12 +234,33 @@ public void start() {
...
+     * @param qryType Query type.
+     * @param isQuery {@code null} if not applicable, {@code true} for select queries,
+     *                           otherwise (DML/DDL queries) {@code false}.
+     */
+    private static void checkQueryType(Type qryType, Boolean isQuery) {
+        if (isQuery == null || qryType == Type.EXPLAIN) {
+            return;
+        }
+        if (isQuery && qryType != Type.QUERY) {
+            throw new IgniteInternalException("Given statement type does not match that declared by JDBC driver");
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:257]: Why does the error that is thrown from common executor module, refers to JDBC driver?

Comment (vladErmakov07) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java]: I refactored error handling and moved all JDBC related messages to JdbcQueryEventHandler. Thnx.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/jdbc/ItJdbcStatementSelfTest.java [hunk 2]
```diff
@@ -718,8 +718,9 @@ public void testStatementTypeMismatchSelectForCachedQuery() throws Exception {
     }
 
     @Test
-    @Disabled("https://issues.apache.org/jira/browse/IGNITE-16268")
     public void testStatementTypeMismatchUpdate() throws Exception {
+        stmt.executeUpdate("update TEST set NAME='28' where ID=1");
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/jdbc/ItJdbcStatementSelfTest.java:722]: this update is hiding original problem

**Review by tledkov-gridgain — APPROVED** [review]
LGTM

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/jdbc/ItJdbcStatementSelfTest.java (hunk 2 above)
Comment (vladErmakov07) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/jdbc/ItJdbcStatementSelfTest.java]: Refactored the test. Test methods are now isolated from each other.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/StatementMismatchException.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.apache.ignite.internal.sql.engine.exec;
+
+import org.apache.ignite.lang.IgniteInternalException;
+
+/**
+ * StatementMismatchException is used during query validation.
+ * The exception is used when the expected query type
+ * does not match the actual query type obtained after parsing a sql string.
+ */
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/StatementMismatchException.java:26]:
```suggestion
/**
 * StatementMismatchException is used during query validation.
 *
 * <p>The exception is used when the expected query type does not match the actual query type obtained after parsing a sql string.
 */
```

**Comment by AMashenkov** [comment]
please, fix the conflicts.

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/query/JdbcStatementType.java [hunk 4]
```diff
@@ -17,29 +17,55 @@
...
-    /** Enumerated values. */
-    private static final JdbcStatementType[] VALS = values();
+    static {
+        STATEMENT_TYPE_IDX = Arrays.stream(values()).collect(
+                Collectors.toMap(JdbcStatementType::getId, Function.identity()));
+    }
 
     /**
-     * Efficiently gets enumerated value from its ordinal.
+     * Gets statement type value from its id.
```
Comment (korlov42) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/query/JdbcStatementType.java:48]:
```suggestion
     * Gets statement type value by its id.
```

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/query/JdbcStatementType.java [hunk 5]
```diff
@@ -17,29 +17,55 @@
...
      *
-     * @param ord Ordinal value.
-     * @return Enumerated value or {@code null} if ordinal out of range.
+     * @param id Ordinal value.
```
Comment (korlov42) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/query/JdbcStatementType.java:50]:
```suggestion
     * @param id The id.
```

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/query/JdbcStatementType.java [hunk 6]
```diff
@@ -17,29 +17,55 @@
...
+     * @return JdbcStatementType value.
+     * @throws IllegalArgumentException If statement is not found.
      */
-    public static JdbcStatementType fromOrdinal(int ord) {
-        return ord >= 0 && ord < VALS.length ? VALS[ord] : null;
+    public static JdbcStatementType getStatement(byte id) {
+        JdbcStatementType value = STATEMENT_TYPE_IDX.get(id);
+
+        checkArgument(value != null, "Unknown jdbcStatementType %s", id);
```
Comment (korlov42) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/query/JdbcStatementType.java:57]: let's use `java.util.Objects#requireNonNull(T, java.lang.String)` instead

**Review by korlov42 — APPROVED** [review]

**Pull request merged**